snuba/state/cache/redis/backend.py

```diff
@@ -0,0 +1,147 @@
+import logging
+import uuid
+from typing import Callable, Optional, TypeVar
+
+from pkg_resources import resource_string
+from redis.exceptions import ResponseError
+
+from snuba.redis import RedisClientType
+from snuba.state import get_config
+from snuba.state.cache.abstract import Cache, ExecutionError, ExecutionTimeoutError
+from snuba.utils.codecs import Codec
+
+
+T = TypeVar("T")
+
+
+logger = logging.getLogger(__name__)
+
+
+RESULT_VALUE = 0
+RESULT_EXECUTE = 1
+RESULT_WAIT = 2
+
+
+class RedisCache(Cache[T]):
+    def __init__(
+        self, client: RedisClientType, prefix: str, codec: Codec[bytes, T]
+    ) -> None:
+        self.__client = client
+        self.__prefix = prefix
+        self.__codec = codec
+
+        # TODO: This should probably be lazily instantiated, rather than
+        # automatically happening at startup.
+        self.__script_get = client.register_script(
+            resource_string("snuba", "state/cache/redis/scripts/get.lua")
+        )
+        self.__script_set = client.register_script(
+            resource_string("snuba", "state/cache/redis/scripts/set.lua")
+        )
+
+    def __build_key(self, key: str) -> str:
+        return f"{self.__prefix}{{{key}}}"
+
+    def get(self, key: str) -> Optional[T]:
+        value = self.__client.get(self.__build_key(key))
+        if value is None:
+            return None
+
+        return self.__codec.decode(value)
+
+    def set(self, key: str, value: T) -> None:
+        self.__client.set(
+            self.__build_key(key),
+            self.__codec.encode(value),
+            ex=get_config("cache_expiry_sec", 1),
+        )
+
+    def get_or_execute(self, key: str, function: Callable[[], T], timeout: int) -> T:
+        result_key = self.__build_key(key)
+        wait_queue_key = f"{self.__build_key(key)}:w"
+        task_ident_key = f"{self.__build_key(key)}:t"
```

Comment:
This keyspace for these items and the existing cache keyspace do overlap however which is potentially a concern that I forgot to note here.

Reply:
About the docstring, I am ok wether some docs go in the docstring or whether they go inside the body of the function. The way that feels more clear to you will be fine.
My point is that this method is really complex. This is real complexity, not incidental, it took two hours to me to assess it was working. I don't know how you feel about that after working on it for a week, but see how long your description of the corner cases was in the comment below. 
I think it would be very useful to have an explanation of the design (not as long as your comment below) that is not scattered around and across different files could help a lot to understand it.

Reply:
I agree. Will improve this. Thanks for the perspective.

Reply:
Documented the algorithm more thoroughly with b689e2d. Hopefully this alleviates some of your concerns.